### PR TITLE
Preserve context values

### PR DESCRIPTION
**Pull request by vvoland**
Instead of passing a fresh `context.Background()` wrap context passed by the user with a context that ignores cancellation.

This allows to drop the original cancellation while preserving context values.

If built with Go 1.21+ the stdlib `context.WithoutCancel` is used.
Otherwise it fallbacks to an in-tree copy of the withoutCancel.

### CODE REVIEW

File: singleflight.go
Comment (vvoland) [inline singleflight.go:33]: This might be regarded as a breaking/behavior change. Wondering if this behavior is desired in the general case, or should this be a separate function (something like `DoWithValues`)?

**Comment by janos** [comment]
Hey, thank you for the PR. On the first glance, it looks great and complete. I would have to review it a bit more before the merge. I must say that I was looking for a way to nicely preserve values since the 0.1 version. Go 1.21 helps a lot there and the backport is clean.

**Comment by thaJeztah** [comment]
Also; thank you for writing this module! :heart: We started using this module in Moby (Docker Engine) 24.0 and up;

- https://github.com/moby/moby/pull/44520

I must admit that I don't recall how we arrived on your module, but @vvoland may remember; it could've been that we started implementing, and discovered "someone already did this!" 😅 

In either case; it's been running on many systems, and as far as I'm aware, no issues so far 🎉 That said, feel free to call out if you need help maintaining the project (e.g. help review pull requests / changes); I'm sure Moby maintainers are happy to help out (feel free to ping any of us in that case 👍)

**Comment by vvoland** [comment]
Ah, it was actually @corhere who found this package and let me know about it in one of my context plumbing PRs: https://github.com/moby/moby/pull/44365#discussion_r1007313872 😄

File: singleflight.go
Comment (janos) [reply singleflight.go:33]: I think that this behaviour is desired in general case, even expected. I would not pollute the package api, just relay on the 0 major version to bump the minor one to reflect the behaviour change.

File: withoutcancel.go
Comment (janos) [reply withoutcancel.go:1]: In this case it looks like needed to avoid double declaration when using different go versions to build.

**Comment by janos** [comment]
I am so glad to read that you are using this package in Moby. :)

There is no big pressure on the maintenance, so far, but a help from the community is always welcome. :)

**Review by janos — APPROVED** [review]
Awesome. :)

File: withoutcancel.go
Comment (vvoland) [reply withoutcancel.go:1]: We still need the explicit go build directive as the file suffixes only work for the GOOS/GOARCH, so it's mostly a matter of convention.
I can rename the `withoutcancel.go` to `withoutcancel_go121.go` (for Go 1.21+) and `withoutcancel_backport.go` to `withoutcancel.go` if you consider it more idiomatic :)

Comment (janos) [reply withoutcancel.go:1]: The rename is more explicit/idiomatic, even cosmetic, I am for it.

Comment (vvoland) [reply withoutcancel.go]: Done!

**Comment by janos** [comment]
Thank you for contributing, resenje.org/singleflight [v0.4.0](https://github.com/janos/singleflight/tree/v0.4.0) is released with this improvement.

**Comment by vvoland** [comment]
No problem, it's nice to contribute back! Thanks for the quick release!

**Pull request merged**
